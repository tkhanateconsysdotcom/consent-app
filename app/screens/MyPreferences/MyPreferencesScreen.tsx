import React, { useCallback, useState } from "react";
import { ScrollView, StyleSheet, View } from "react-native";
import { useDispatch, useSelector } from "react-redux";
import uuid from "uuid";

import { Button, ListItem, Text, TextInput } from "components";
import { userPreferences } from "ducks";
import { Layout } from "constant";
import { Preference } from "interfaces";
import { PreferenceSelection } from "components/PreferenceSelection/PreferenceSelection";

import navigationOptions from "./MyPreferencesScreen.navigationOptions";
import styles from "./MyPreferencesScreen.styles";

interface Props {
  preferenceString: string;
}

const AddPreferenceButton = ({ preferenceString }: Props) => {
  const dispatch = useDispatch();

  const addPreference = () => {
    const preference = {
      preference: preferenceString,
      id: uuid(),
      isMitigated: false,
    };

    dispatch(userPreferences.actions.addPreference(preference));
  };

  const styles = StyleSheet.create({
    button: {
      margin: Layout.PADDING_HORIZONTAL,
    },
  });

  return (
    <Button.Primary
      onPress={addPreference}
      textType={"Primary"}
      style={styles.button}
    >
      <Text.Primary white={false} red={true} center bold>
        {"Add Preference"}
      </Text.Primary>
    </Button.Primary>
  );
};

const PARTNER_NAME_MAX_LENGTH = 150;

const MyPreferencesScreen = () => {
  const [preferenceName, setPartnerName] = useState<string>("");
  const onChangePartnerName = useCallback((name: string) => {
    if (name.length < PARTNER_NAME_MAX_LENGTH) {
      setPartnerName(name);
    }
  }, []);
  const dispatch = useDispatch();
  const preferencesSelector: Preference[] = useSelector(
    userPreferences.selectors.getPreference
  );
  const preferences = preferencesSelector ? preferencesSelector : [];

  const onPress = useCallback(
    (pref: Preference) => {
      dispatch(userPreferences.actions.removePreference(pref));
    },
    [dispatch]
  );

  function foo() {
    // do nothing.
  }

  return (
    <View style={styles.container}>
      <Text.Primary style={styles.intro}>
        {"Add any actions you will allow with your partners:"}
      </Text.Primary>
      <View style={styles.acceptanceIntensityContainer}>
        <TextInput
          isOptional={false}
          placeholder={"Acceptable Actions"}
          title={"Please add any preferences you have"}
          onChangeText={onChangePartnerName}
          value={preferenceName}
        />
      </View>
      <View>
        <AddPreferenceButton preferenceString={preferenceName} />
      </View>
      <ScrollView style={styles.scrollContainer}>
        <ListItem title={"SELECT A PREFERENCE TO REMOVE"} onPress={foo} />
        {!Array.isArray(preferences)
          ? () => {
              // do nothing
            }
          : preferences.map((p: Preference) => (
              <PreferenceSelection
                key={p.id}
                preference={p}
                onSelectPreference={() => onPress(p)}
                selectedPreference={p}
              />
            ))}
      </ScrollView>
    </View>
  );
};

MyPreferencesScreen.navigationOptions = navigationOptions;

export default MyPreferencesScreen;
