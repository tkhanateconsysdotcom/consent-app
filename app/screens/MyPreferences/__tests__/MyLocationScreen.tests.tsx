import React from "react";
import { create } from "react-test-renderer";

import MyPreferencesScreen from "../MyPreferencesScreen";

// jest.unmock("../");

it("MyPreferencesScreen renders correctly", () => {
  const tree = create(<MyPreferencesScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});
