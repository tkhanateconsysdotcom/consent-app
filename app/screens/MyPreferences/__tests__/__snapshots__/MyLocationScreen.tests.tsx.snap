// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyPreferencesScreen renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "paddingHorizontal": 14,
    }
  }
>
  <Text.Primary
    style={
      Object {
        "paddingTop": 30,
      }
    }
  >
    Add any actions you will allow with your partners:
  </Text.Primary>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <TextInput
      isOptional={false}
      onChangeText={[Function]}
      placeholder="Acceptable Actions"
      title="Please add any preferences you have"
      value=""
    />
  </View>
  <View>
    <Button.Primary
      onPress={[Function]}
      style={
        Object {
          "margin": 14,
        }
      }
      textType="Primary"
    >
      <Text.Primary
        bold={true}
        center={true}
        red={true}
        white={false}
      >
        Add Preference
      </Text.Primary>
    </Button.Primary>
  </View>
  <RCTScrollView
    style={
      Object {
        "borderTopColor": "#DCF0E5",
        "borderTopWidth": 2,
      }
    }
  >
    <View>
      <TouchableOpacity
        onPress={[Function]}
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <Text.Secondary
          numberOfLines={1}
          style={
            Object {
              "paddingVertical": 20,
            }
          }
        >
          SELECT A PREFERENCE TO REMOVE
        </Text.Secondary>
        <Ionicons
          color="#9CA6A0"
          name="ios-chevron-forward-outline"
          size={26}
          style={
            Object {
              "marginTop": 18,
            }
          }
        />
      </TouchableOpacity>
    </View>
  </RCTScrollView>
</View>
`;
